### PR TITLE
Feature: Don't reorder files after renaming them

### LINKED ISSUE #4214: Item is not moving after changing name
**Is your feature request related to a problem?**
A clear description of what the problem is.

**Describe the solution**
A clear description of what you want.

**Describe alternatives you've considered**
A clear description of any alternative solutions or features you've considered.

**Additional comment**
Add any other comment or screenshots about the feature request here. What device type/UI layout and app version?

### PR DESCRIPTION
**Pull request by Spid3rrr**
This adds the feature that files should not be re-ordered after renaming them.

**Resolved / Related Issues**
- [x] Were these changes approved in an issue or discussion with the project maintainers? In order to prevent extra work, feature requests and changes to the codebase must be approved before the pull request will be reviewed. This prevents extra work for the contributors and maintainers.
   Closes #4214 

**Validation**
How did you test these changes?
- [x] Did you build the app and test your changes?
- [ ] Did you check for accessibility? You can use Accessibility Insights for this.
- [ ] Did you remove any strings from the en-us resource file?
   - [ ] Did you search the solution to see if the string is still being used? 
- [ ] Did you implement any design changes to an existing feature?
   - [ ] Was this change approved?
- [ ] Are there any other steps that were used to validate these changes?

**Screenshots (optional)**
Add screenshots here.

### CODE REVIEW

**Comment by yaira2** [comment]
@Spid3rrr the CI automation tests are failing, can you look into this?

File: src/Files.App/ViewModels/ItemViewModel.cs [hunk 1]
```diff
@@ -1973,6 +1972,11 @@ async Task HandleChangesOccurredAsync()
 					{
 						operationEvent.Reset();
 
+						// Keep track of renaming information available
+						int RenameIndex = -1;
+						string OldName = "";
+						string NewName = "";
```
Comment (ferrariofilippo) [inline src/Files.App/ViewModels/ItemViewModel.cs:1978]:
```suggestion
						int renameIndex = -1;
						string oldName = string.Empty;
						string newName = string.Empty;
```

You should use lowercase names for variable declarations. I think that using `string.Empty` will make the code more readable than `""`

File: src/Files.App/ViewModels/ItemViewModel.cs [hunk 2]
```diff
@@ -2012,6 +2018,16 @@ async Task HandleChangesOccurredAsync()
 								App.Logger.LogWarning(ex, ex.Message);
 							}
 
+							if (RenameIndex != -1 && OldName != "" && NewName != "")
+							{
+								var itemRemoved = await RemoveFileOrFolderAsync(OldName);
+								var itemAdded = await AddFileOrFolderAsync(NewName, RenameIndex);
```
Comment (ferrariofilippo) [inline src/Files.App/ViewModels/ItemViewModel.cs:2024]: I don't think you need to assign these two variables, you never read them

```suggestion
								await RemoveFileOrFolderAsync(OldName);
								await AddFileOrFolderAsync(NewName, RenameIndex);
```

File: src/Files.App/ViewModels/ItemViewModel.cs [hunk 3]
```diff
@@ -2012,6 +2018,16 @@ async Task HandleChangesOccurredAsync()
 								App.Logger.LogWarning(ex, ex.Message);
 							}
 
+							if (RenameIndex != -1 && OldName != "" && NewName != "")
```
Comment (ferrariofilippo) [inline src/Files.App/ViewModels/ItemViewModel.cs:2021]: I'd use `string.Empty` here, as mentioned above

File: src/Files.App/ViewModels/ItemViewModel.cs [hunk 4]
```diff
@@ -2012,6 +2018,16 @@ async Task HandleChangesOccurredAsync()
...
+							{
+								var itemRemoved = await RemoveFileOrFolderAsync(OldName);
+								var itemAdded = await AddFileOrFolderAsync(NewName, RenameIndex);
+								RenameIndex = -1;
+								OldName = "";
+								NewName = "";
```
Comment (ferrariofilippo) [inline src/Files.App/ViewModels/ItemViewModel.cs:2027]: Same `string.Empty` thing

File: src/Files.App/ViewModels/ItemViewModel.cs [hunk 5]
```diff
@@ -2084,7 +2100,45 @@ private async Task AddFileOrFolderAsync(ListedItem? item)
...
+			{
+				if (index == -1)
+				{
+					filesAndFolders.Add(item);
+				}
+				else
+				{
+					filesAndFolders.Insert(index, item);
+				}
+				if (UserSettingsService.FoldersSettingsService.AreAlternateStreamsVisible)
```
Comment (ferrariofilippo) [inline src/Files.App/ViewModels/ItemViewModel.cs:2127]:
```suggestion
				if (index == -1)
					filesAndFolders.Add(item);
				else
					filesAndFolders.Insert(index, item);

				if (UserSettingsService.FoldersSettingsService.AreAlternateStreamsVisible)
```

You should avoid using `{}` for single line `ifs` (see <a href="https://files.community/docs/contributing/code-style">style guidelines</a>). Also, adding a new line to split the two ifs will make the code more readable

**Review by ferrariofilippo — COMMENTED** [review]
The feature itself works, but I think there are some little tweaks which may improve code readability

File: src/Files.App/ViewModels/ItemViewModel.cs [hunk 6]
```diff
@@ -2119,7 +2173,14 @@ private async Task AddFileOrFolderAsync(ListedItem? item)
...
 
-			await AddFileOrFolderAsync(listedItem);
+			if (index == -1)
+			{
+				await AddFileOrFolderAsync(listedItem);
+			}
+			else
+			{
+				await AddFileOrFolderAsync(listedItem, index);
+			}
```
[the remark this answers is not in the document]
Comment (Spid3rrr) [reply src/Files.App/ViewModels/ItemViewModel.cs:2183]: Oh I see ! That would be a lot better, I don't know why it slipped me I'm doing the exact same thing !

**Comment by Spid3rrr** [comment]
@ferrariofilippo I implemented the code suggestions in the latest commit. I am still figuring out why the automated testing is failing. I believe it may have been testing the previous behavior (renaming a file and trying to locate it at the bottom of the tab ?) but I'm still investigating.

**Comment by Spid3rrr** [comment]
Update : The recent for the check failing was that the folder/file wasn't being selected after renaming. That should be fixed now.

File: src/Files.App/ViewModels/ItemViewModel.cs [hunk 7]
```diff
@@ -2068,7 +2084,14 @@ private async Task AddFileOrFolderAsync(ListedItem? item)
...
 			{
-				filesAndFolders.Add(item);
+				if (index == -1)
+
+					filesAndFolders.Add(item);
+
+				else
+
+					filesAndFolders.Insert(index, item);
+
```
Comment (ferrariofilippo) [inline src/Files.App/ViewModels/ItemViewModel.cs:2094]:
```suggestion
				if (index == -1)
					filesAndFolders.Add(item);
				else
					filesAndFolders.Insert(index, item);
```

**Review by ferrariofilippo — DISMISSED** [review]
LGTM and works for me!
(There's a small thing with code style)

**Comment by yaira2** [comment]
@Spid3rrr how does this work if the user renames multiple files simultaneously (through an app like PowerToys or File Explorer)?

**Comment by Spid3rrr** [comment]
@yair100 the files renamed are being re-ordered even from this pull request ?

**Comment by yaira2** [comment]
Yes

**Comment by yaira2** [comment]
@Spid3rrr I did some more testing and confirmed this PR does work in regular folders, however the issue is still present if you use a grouping option (eg group by date).

**Comment by Spid3rrr** [comment]
Oh I understand. I will look into it.
________________________________

**Comment by yaira2** [comment]
Thanks!

**Comment by yaira2** [comment]
@Spid3rrr

**Comment by yaira2** [comment]
@Spid3rrr

**Comment by Spid3rrr** [comment]
I apologize I haven't had time to look at it yet. I will update this week !

**Pull request closed without merging**